v5/core/iam_authenticator.go

```diff
@@ -165,23 +240,40 @@ func (authenticator *IamAuthenticator) setTokenData(tokenData *iamTokenData) {
 	defer authenticator.tokenDataMutex.Unlock()
 
 	authenticator.tokenData = tokenData
+
+	// Next, we should save the just-returned refresh token back to the main
+	// authenticator struct.
+	// This is done so that if we were originally configured with
+	// a refresh token, then we'll be sure to use a "fresh"
+	// refresh token next time we invoke the "get token" operation.
+	// This was recommended by the IAM team to avoid problems in the future
+	// if the token service is changed to invalidate an existing refresh token
+	// when a new one is generated and returned in the response.
+	if tokenData != nil {
+		authenticator.RefreshToken = tokenData.RefreshToken
+	}
 }
 
 // Validate the authenticator's configuration.
 //
-// Ensures the ApiKey is valid, and the ClientId and ClientSecret pair are
-// mutually inclusive.
+// Ensures that the ApiKey and RefreshToken properties are mutually exclusive,
+// and that the ClientId and ClientSecret properties are mutually inclusive.
 func (this *IamAuthenticator) Validate() error {
-	if this.ApiKey == "" {
-		return fmt.Errorf(ERRORMSG_PROP_MISSING, "ApiKey")
+
+	// The user should specify exactly one of ApiKey or RefreshToken.
+	if this.ApiKey == "" && this.RefreshToken == "" ||
+		this.ApiKey != "" && this.RefreshToken != "" {
+		return fmt.Errorf(ERRORMSG_EXCLUSIVE_PROPS_ERROR, "ApiKey", "RefreshToken")
 	}
 
-	if HasBadFirstOrLastChar(this.ApiKey) {
+	if this.ApiKey != "" && HasBadFirstOrLastChar(this.ApiKey) {
 		return fmt.Errorf(ERRORMSG_PROP_INVALID, "ApiKey")
 	}
 
-	// Validate ClientId and ClientSecret.  They must both be specified togther or neither should be specified.
-	if this.ClientId == "" && this.ClientSecret == "" {
+	// Validate ClientId and ClientSecret.
+	// If RefreshToken is not specified, then both or neither should be specified.
+	// If RefreshToken is specified, then both must be specified.
+	if this.ClientId == "" && this.ClientSecret == "" && this.RefreshToken == "" {
```

Comment:
Is it possible that the API key was created without either the client ID or secret set? I guess there's a default that's used in that case but would the refresh token path use the default as well?

Reply:
Great questions...
When obtaining an access token for an apikey, there are no "default" values for client id/secret that are used if they're not specified. In this scenario, you won't get back a valid refresh token.

In order to obtain a valid refresh token, you must specify "bx/bx" for the client id/secret values along with the apikey.
So, if this refresh token is then going to be used to construct the IAM Authenticator, the same client id/secret values must be used as well.   I didn't want to hardcode client id/secret to be "bx/bx" in the authenticator when a refresh token is being used because that seems a bit hackish.  I discussed this with Martin Smolny and we settled on "the client id/secret values must be the same values that were used to initially obtain the refresh token", even though in the IBM IAM implementation those values an only be "bx/bx".  However, in the future, this constraint might be relaxed to support additional values for those properties, but the "same as" constraint would still apply... presumably :)